pyop2/types/dat.py

```diff
@@ -539,14 +540,34 @@ def __itruediv__(self, other):
         """Pointwise division or scaling of fields."""
         return self._iop(other, operator.itruediv)
 
+    @mpi.collective
+    def freeze_halo(self):
+        """Disable halo exchanges.
+
+        This is useful in cases where one is repeatedly writing to (and not reading from)
+        a :class:`Dat` as the intermediate updates can be skipped.
+
+        Note that the user is responsible for ensuring that the :class:`Dat` is in a
+        consistent state prior to freezing (by calling :func:`Dat.global_to_local_begin`
+        and :func:`Dat.global_to_local_end`) and also after unfreezing (with
+        :func:`Dat.local_to_global_begin` and :func:`Dat.local_to_global_end`).
```

Comment:
This sounds like it's screaming out for a context manager which does this setup and unsetup for you. I.e.:
```python3
with dat.frozen_halo():
     assemble(...)
     assemble(...)
     assemble(...)
```

Reply:
You're quite right. I'll try to get that working.

Reply:
I would advocate a model:

```python
def freeze_with(dat):
    dat.global_to_local_begin(op2.INC)
    dat.global_to_local_end(op2.INC)

def unfreeze_with(dat):
   ...

with dat.frozen_halo(pre=freeze_with, post=unfreeze_with):
    ...
```

Just FYI, don't use `contextlib.contextmanager` to make your context manager (since it is quite a bit slower than writing a class that provides a context manager interface by hand).

Reply:
I wish I had read this 20 minutes ago...

Reply:
Arguably the performance difference is not worth the effort of writing something more complicated. The context manager is supposed to live outside parloops so there's no way this is ever a bottleneck.

Reply:
FWIW, the "more complicated thing" is:
```python
class frozen_halo:
    def __init__(self, pre, post):
        self.pre = pre
        self.post = post
    def __enter__(self):
         self.pre()
    def __exit__(self, exc_typ, exc_val, tb):
         # handle exception if you like
         self.post()

class Dat:
     def frozen_halo(self, access):
          return frozen_halo(...)
```

Reply:
Thanks. I'll have a go at implementing this.